### PR TITLE
Memory fixes

### PR DESCRIPTION
**Pull request by gperciva**
For tarsnap_mode_c(): We need two independent clean-up paths in the code.  Other than moving code into the

    if (failure) {
        ...
    }

parts directly (which might not be terrible for the ccache_write() failure), I couldn't see a cleaner way of doing this.  However, I realize that two distinct clean-up paths is not usual for tarsnap, and the error labels might need renaming.  I didn't find any inspiration from looking at other `goto` labels in tarsnap.

### CODE REVIEW

File: lib/netproto/netproto_connect.c [hunk 1]
```diff
@@ -105,11 +116,16 @@ callback_cancel(void * cookie)
 static struct sock_addr **
 getserveraddr(void)
 {
-	static struct sock_addr ** srv_addr = NULL;
 	static time_t srv_time = (time_t)(-1);
 	struct sock_addr ** tmp_addr;
 	time_t tmp_time;
 
+	/* This should only happen once. */
```
Comment (cperciva) [inline lib/netproto/netproto_connect.c]: To make the code simpler, how about moving this...

**Comment by cperciva** [comment]
Can you split the write.c bit off into a separate PR?  It's less obvious and I want to review it more later.

File: lib/netproto/netproto_connect.c (hunk 1 above)
Comment (gperciva) [inline lib/netproto/netproto_connect.c]: I wasn't certain if there was any multi-threading issues here.  I mean, how long can `sock_resolve` take?  with the callbacks, I wasn't certain if you might call `getserveraddr` multiple times (before the first had finished).

However, I see that you don't use `pthread.h`, so I'll go ahead and make this change.  :)

**Comment by gperciva** [comment]
Most of these commits were already accepted; I'm working on a better version of the final one.

**Pull request closed without merging**